debezium-connector-mongodb/src/main/java/io/debezium/connector/mongodb/MongoDbIncrementalSnapshotChangeEventSource.java

```diff
@@ -0,0 +1,452 @@
+    }
+
+    protected String getSignalCollectionName(String dataCollectionId) {
+        return dataCollectionId;
+    }
+
+    protected void sendWindowEvents(Partition partition, OffsetContext offsetContext) throws InterruptedException {
+        LOGGER.debug("Sending {} events from window buffer", window.size());
+        offsetContext.incrementalSnapshotEvents();
+        for (Object[] row : window.values()) {
+            sendEvent(partition, dispatcher, offsetContext, row);
+        }
+        offsetContext.postSnapshotCompletion();
+        window.clear();
+    }
+
+    // TODO Used typed dispatcher and offset context
+    protected void sendEvent(Partition partition, EventDispatcher dispatcher, OffsetContext offsetContext, Object[] row) throws InterruptedException {
+        context.sendEvent(keyFromRow(row));
+        ((ReplicaSetOffsetContext) offsetContext).readEvent((CollectionId) context.currentDataCollectionId(), clock.currentTimeAsInstant());
+        dispatcher.dispatchSnapshotEvent(context.currentDataCollectionId(),
+                getChangeRecordEmitter(partition, context.currentDataCollectionId(), offsetContext, row),
+                dispatcher.getIncrementalSnapshotChangeEventReceiver(dataListener));
+    }
+
+    /**
+     * Returns a {@link ChangeRecordEmitter} producing the change records for
+     * the given table row.
+     */
+    protected ChangeRecordEmitter getChangeRecordEmitter(Partition partition, T dataCollectionId,
+                                                         OffsetContext offsetContext, Object[] row) {
+        return new MongoDbChangeSnapshotOplogRecordEmitter(partition, offsetContext, clock, (Document) row[0], true);
+    }
+
+    protected void deduplicateWindow(DataCollectionId dataCollectionId, Object key) {
+        if (!context.currentDataCollectionId().equals(dataCollectionId)) {
+            return;
+        }
+        if (key instanceof Struct) {
+            if (window.remove((Struct) key) != null) {
+                LOGGER.info("Removed '{}' from window", key);
+            }
+        }
+    }
+
+    /**
+     * Update low watermark for the incremental snapshot chunk
+     */
+    protected void emitWindowOpen() throws InterruptedException {
+        final CollectionId collectionId = signallingCollectionId;
+        final String id = context.currentChunkId() + "-open";
+        primary.executeBlocking(
+                "emit window open for chunk '" + context.currentChunkId() + "'",
+                primary -> {
+                    final MongoDatabase database = primary.getDatabase(collectionId.dbName());
+                    final MongoCollection<Document> collection = database.getCollection(collectionId.name());
+
+                    LOGGER.trace("Emitting open window for chunk = '{}'", context.currentChunkId());
+                    collection.insertOne(Document.parse("{\"_id\": \"" + id + "\", \"type\": \"" + OpenIncrementalSnapshotWindow.NAME + "\", \"payload\": \"\"}"));
```

Comment:
Let's use the programmatic API for assembling a doc instead of that string-parsing hack.